pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -82,37 +82,40 @@ public class ModelInterpreter implements Serializable {
-                        // If we have an agent and script.scm isn't null, run checkout scm
-                        if (root.agent.hasAgent() && Utils.hasScmContext(script)) {
-                                script.checkout script.scm
+                    // Stage execution and post-build actions run in try/catch blocks, so we still run post-build actions
+                    // even if the build fails, and we still send notifications if the build and/or post-build actions fail.
+                    // We save the caught error, if any, for throwing at the end of the build.
+                    inDeclarativeAgent(root.agent) {
+                        toolsBlock(root.agent, root.tools) {
+                            // If we have an agent and script.scm isn't null, run checkout scm
+                            if (root.agent.hasAgent() && Utils.hasScmContext(script)) {
+                                script.stage(SyntheticStage.checkout()) {
+                                    Utils.markSyntheticStage(SyntheticStage.checkout(), SyntheticStage.SYNTHETIC_PRE)
+                                    script.checkout script.scm
+                                }
                             }
 
                             for (int i = 0; i < root.stages.getStages().size(); i++) {
                                 Stage thisStage = root.stages.getStages().get(i)
 
-                            script.stage(thisStage.name) {
-                                withEnvBlock(thisStage.getEnvVars()) {
-                                    if (firstError == null) {
-                                        inDeclarativeAgent(thisStage.agent) {
-                                            toolsBlock(thisStage.agent ?: root.agent, thisStage.tools) {
-                                                try {
-                                                    catchRequiredContextForNode(root.agent) {
-                                                        setUpDelegate(thisStage.steps.closure).call()
-                                                    }.call()
-                                                } catch (Exception e) {
-                                                    script.echo "Error in stages execution: ${e.getMessage()}"
-                                                    script.getProperty("currentBuild").result = Result.FAILURE
-                                                    if (firstError == null) {
-                                                        firstError = e
-                                                    }
-                                                } finally {
-                                                    // And finally, run the post stage steps.
-                                                    List<Closure> postClosures = thisStage.satisfiedPostStageConditions(root, script.getProperty("currentBuild"))
+                                script.stage(thisStage.name) {
```

Comment:
Because I'm fixing some indent problems while I'm here so as not to get so hateful at everything. =)